File: fnc/examples/chapter13/wave2D.m
function wave2D    % ignore this line

%%
m = 60;  n = 60;
[X,Y,d] = rectdisc(m,[-2,2],n,[-2,2]);

%%
% Here is the initial condition. The boundary values of $u$ will remain
% constant.
U0 = (X+.2).*exp(-12*(X.^2+Y.^2));
V0 = 0*U0;
pcolor(X,Y,U0)
caxis([-.12 .12]), axis equal, shading flat    % ignore this line
title('Initial condition')    % ignore this line
xlabel('x'), ylabel('y')    % ignore this line

%%
% The |unpack| function separates the unknowns for $u$ and $v$, applies
% the boundary conditions on $u$, and returns two functions on the grid.
    function [U,V] = unpack(w)
        numU = (m-1)*(n-1);  % number of unknowns for U
        U = U0;               
        U(~d.isbndy) = w(1:numU); % overwrite the interior
        V = d.unvec( w(numU+1:end) );     % use all values
    end

%%
% The next function drops the boundary values of $u$ and returns a vector
% of all the unknowns for both components of the solution. It's the inverse
% of the |unpack| function.
    function w = pack(U,V)
        w = U(~d.isbndy);
        w = [ w; V(:) ];
    end

%%
% The following function computes the time derivative of the unknowns. Besides the
% translation between vector and matrix shapes, it's quite straightforward.
    function dwdt = timederiv(t,w)
        [U,V] = unpack(w); 
        dUdt = V;
        dWdt = d.Dxx*U + U*d.Dyy';
        dwdt = pack(dUdt,dWdt);
    end

%%
% Since this problem is hyperbolic, not parabolic, a nonstiff integrator
% like |ode45| is fine and faster than a stiff integrator.
t = linspace(0,3,5);
v0 = pack(U0,V0);
[t,W] = ode45(@timederiv,t,v0);
W = W';  % each column is one time instant

%%
% We plot the solution at four different times. 
for k = 1:2
    subplot(1,2,k)
    [U,~] = unpack(W(:,k+1));
    pcolor(X,Y,U)
    axis equal, shading flat     % ignore this line
    caxis([-.1 .1])     % ignore this line
    title(sprintf('t = %.2f',t(k+1)))     % ignore this line
    xlabel('x'), ylabel('y')    % ignore this line
end

%%
for k = 1:2
    subplot(1,2,k)
    [U,~] = unpack(W(:,k+3));
    pcolor(X,Y,U)
    axis equal, shading flat     % ignore this line
    caxis([-.1 .1])     % ignore this line
    title(sprintf('t = %.2f',t(k+3)))     % ignore this line
    xlabel('x'), ylabel('y')    % ignore this line
end

end    % ignore this line

File: julia/scripts/chapter13/wave2D.jl

include("../FNC.jl")
m = 60;  n = 60;
X,Y,d = FNC.rectdisc(m,[-2,2],n,[-2,2]);
x = X[:,1]; y = Y[1,:];  # for plotting

U0 = @. (X+0.2)*exp(-12*(X^2+Y^2))
V0 = zeros(size(U0))

using Plots
surface(x,y,U0,     
    title="Initial condition",    
    xaxis=("x"),yaxis=("y") ) 

unpack = function (w)
    numU = (m-1)*(n-1)   # number of unknowns for U
    U = copy(U0)               
    U[@. !d.isbndy] = w[1:numU]  # overwrite the interior
    V = d.unvec( w[numU+1:end] )     # use all values
    return U,V
end

pack = function (U,V)
    w = U[@. !d.isbndy]
    return [ w; V[:] ]
end

dwdt = function (w,tmp,t)
    U,V = unpack(w) 
    dUdt = V
    dWdt = d.Dxx*U + U*d.Dyy'
    return pack(dUdt,dWdt)
end

using DifferentialEquations
IVP = ODEProblem(dwdt,pack(U0,V0),(0,3.0))
sol = solve(IVP,alg_hints=[:nonstiff]);

an = @animate for t = range(0,stop=3,length=80)
    surface(x,y,unpack(sol(t)),match_dimensions=true,color=:redsblues,clims=(-0.1,0.1),
        title="Wave equation solution at t=$(round(t,digits=2))")
end
gif(an,"wave2D.gif")
